internal/handlers/pg/pgdb/pool.go

```diff
@@ -413,3 +440,97 @@ func (pgPool *Pool) SchemaStats(ctx context.Context, schema, collection string)
 	}
 	return &res, nil
 }
+
+// CreateSettingsTable creates FerretDB settings table.
+func (pgPool *Pool) CreateSettingsTable(ctx context.Context, db string) error {
+	sql := `CREATE TABLE ` + pgx.Identifier{db, collectionPrefix + "settings"}.Sanitize() + ` (_jsonb jsonb)`
+	_, err := pgPool.Exec(ctx, sql)
+	if err != nil {
+		return err
+	}
+
+	settings := must.NotFail(types.NewDocument("collections", must.NotFail(types.NewDocument())))
+	sql = fmt.Sprintf(`INSERT INTO %s (_jsonb) VALUES ($1)`, pgx.Identifier{db, collectionPrefix + "settings"}.Sanitize())
+	_, err = pgPool.Exec(ctx, sql, must.NotFail(fjson.Marshal(settings)))
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	return nil
+}
+
+func (pgPool *Pool) GetTableName(ctx context.Context, db, collection string) (string, error) {
+	tableExists, err := pgPool.TableExists(ctx, db, collectionPrefix+"settings")
+	if err != nil {
+		return "", lazyerrors.Error(err)
+	}
+	if !tableExists {
+		err = pgPool.CreateSettingsTable(ctx, db)
+		if err != nil {
+			return "", lazyerrors.Error(err)
+		}
+	}
+
+	sql := `SELECT _jsonb FROM ` + pgx.Identifier{db, collectionPrefix + "settings"}.Sanitize()
+	rows, err := pgPool.Query(ctx, sql)
+	if err != nil {
+		return "", lazyerrors.Error(err)
+	}
+
+	if !rows.Next() {
+		return "", fmt.Errorf("no rows returned")
+	}
+
+	var b []byte
+	if err := rows.Scan(&b); err != nil {
+		return "", lazyerrors.Error(err)
+	}
+
+	doc, err := fjson.Unmarshal(b)
+	if err != nil {
+		return "", lazyerrors.Error(err)
+	}
+
+	settings, err := common.AssertType[*types.Document](doc)
+	if err != nil {
+		return "", lazyerrors.Error(err)
+	}
+
+	collections, err := common.AssertType[*types.Document](must.NotFail(settings.Get("collections")))
+	if err != nil {
+		return "", lazyerrors.Error(err)
+	}
+
+	if collections.Has(collection) {
+		return must.NotFail(collections.Get(collection)).(string), nil
+	}
+
+	tableName := must.NotFail(getTableNameFormatted(collection))
+	must.NoError(collections.Set(collection, tableName))
+	must.NoError(settings.Set("collections", collections))
+
+	sql = `UPDATE ` + pgx.Identifier{db, collectionPrefix + "settings"}.Sanitize() + `SET _jsonb = $1`
```

Comment:
Yeah, this is what I mean. It looks quite unsafe from the race condition perspective.